switch to C++20

psychocoderHPC opened the pull request:
- update docuemntation
- update CI tests
  - remove HIP 5.X
  - remove CUDA 11.X
  - remove GCC9
- set in all CMake files the C++ standard to C++20
-  update to [mallocMC 2.7.0-dev](https://github.com/alpaka-group/mallocMC/tree/127b29fabbb01c9a055bb5d21213daf2db5e6843)
  - This update is required to remove a deprecated warning triggered when switching to C++20 because most volatile usages are deprecated with C++20.

  ```
  GIT_AUTHOR_NAME="Third Party" GIT_AUTHOR_EMAIL="picongpu@hzdr.de" \
    git subtree pull --prefix thirdParty/mallocMC \
    git@github.com:ComputationalRadiationPhysics/mallocMC.git dev --squash
  rm -r thirdParty/mallocMC/alpaka
  git merge --continue
  ```
- CI: **disable HIP** checks
  - the HIP container using a to old libstdc++ and base compiler which is not C++20 compatible
  - using clangs `libc++` is not possible because it is not shipped with the clang compiler provided by AMD's ROCM
  - the checks will be re-enabled after new base container are available

psychocoderHPC committed:
switch to C++

- update docuemntation
- update CI tests
  - remove HIP 5.X
  - remove CUDA 11.X
  - remove GCC9
- set in all CMake files the C++ standard to C++20

ikbuibui reviewed (dismissed):
This line in the readme needs to be updated
`[![Language](https://img.shields.io/badge/language-C%2B%2B17-orange.svg)](https://isocpp.org/)`

psychocoderHPC committed:
change required CMake to 3.25+

ikbuibui commented on thirdParty/mallocMC/src/include/mallocMC/allocator.hpp:
```diff
@@ -140,7 +140,7 @@ namespace mallocMC
          * register usage in CUDA.
          */
-        alloc(AlpakaDevice& dev, AlpakaQueue& queue, volatile size_t size)
+        alloc(AlpakaDevice& dev, AlpakaQueue& queue, size_t size)
```
is this safe? The comment above indicates that the volatile is really required.

psychocoderHPC replied:
good catch!
no it is not safe, therefor I set the PR to draft, I was testing first if this will solve the Ci issues. I will prepare a PR to update mallocMC to latest dev where the volatile is removed. Volatile was a workaround for a workaround which is most likely not required anymore, therefore it is already removed upstream.

psychocoderHPC committed:
fix CI compile issues

- `boost::call_traits` run into compile issues when using clang 18 and cuda 12.1 and boost 1.77
- fix `CONST_VECTOR`
  - for clang CUDA the guard macros was wrongly set

ikbuibui commented on .gitlab-ci.yml:
```diff
@@ -9,6 +9,7 @@ stages:
 variables:
   CONTAINER_TAG: "3.2"
   PIC_CI_MAINLINE: "TRUE"
+  GIT_SUBMODULE_UPDATE_FLAGS: normal
```
```suggestion
  GIT_SUBMODULE_STRATEGY: normal
```

chillenzer commented:
A pertinent warning:
```
/home/lenz93/workspace/src/picongpu/include/picongpu/../pmacc/math/vector/Vector.hpp: In lambda function:
/home/lenz93/workspace/src/picongpu/include/picongpu/../pmacc/math/vector/Vector.hpp:188:18: warning: implicit capture of 'this' via '[=]' is deprecated in C++20 [-Wdeprecated]
  188 |                 return Vector([=](uint32_t const i) constexpr { return -(*this)[i]; });
      |                  ^
/home/lenz93/workspace/src/picongpu/include/picongpu/../pmacc/math/vector/Vector.hpp:188:18: note: add explicit 'this' or '*this' capture
```

Commit:
Squashed 'thirdParty/mallocMC/' changes from 89c9597c91..8624e978e8

8624e978e8 Merge pull request #276 from chillenzer/fix-submodules
209ac2ff50 Fix submodules
127b29fabb Merge pull request #271 from chillenzer/topic-flatterScatter
7846a36fc7 Add submodule to CI
42a96e94ae Remove redundant memset
36e91a9dcf Update license headers
660ad4fffd Minor adjustments to CI
7665162fa1 Update CMakeLists.txt including version bump for Catch2
10221ebded Add catch2 as submodule
8115fc93c6 Update tests
8485629cf2 Update examples
4d0b90b6fc Update LICENSE, README and Usage
745cb2a333 Add FlatterScatter
0bdd7b5117 Run clang-format
2596432b50 Update utils
e8529c4575 Update XMallocSIMD
d752602f19 Update old creation policies
4058862dcc Update device_allocator.hpp
b48b750431 Update allocator.hpp
76c4f46a61 Merge pull request #270 from chillenzer/update-before-flatterscatter
363090535e Merge pull request #269 from chillenzer/update-zenodo
16c5d0de5b Run clang-format
61154b1098 Update utils
0b69f038ff Update XMallocSIMD
79c1962f99 Update old creation policies
4a6a444911 Update device_allocator.hpp
2a9e806119 Update allocator.hpp
1fc71c5294 Update .zenodo.json
f6af760bd7 Merge pull request #268 from chillenzer/update-tooling
8c897de3d4 Merge pull request #267 from chillenzer/update-clang-format
8a1281a0bb Update .gitignore
14ef1ec4c0 Minor tweaks in github action
e7450d9f44 Clean and add to .clang-tidy
fbefa18015 Run clang-format
1fea13bb87 Copy .clang-format from alpaka
29d275791d Merge pull request #266 from chillenzer/add-pre-commit
bd28b52b32 Remove outdated enforce-ascii from pre-commit config
66a7d41914 Add pre-commit lite action
85a1e480b1 Run pre-commit
ff530bb0b5 Add .pre-commit-config.yaml
486a2956d0 Merge pull request #264 from chillenzer/bump-C++20
602e95db5f Bump C++ version to 20
0b9231a754 Merge pull request #263 from chillenzer/alpaka-1.2.0
fa52618c4f Run clang-format
c9c5ae94e4 Bump version of clang-format, its action and checkout
25f42e3aea Update alpaka version number in INSTALL.md
1e6f75bbf8 Merge commit '30ee4f3ab8d7e42ef20cb5495cc3a94b15193abb' as 'alpaka'
30ee4f3ab8 Squashed 'alpaka/' content from commit fe9436d5
4513a7a81c Remove alpaka to pull the latest release
2689825301 LICENSE file remove C-comment (#262)
7bf50bf587 Bump version to 3.0.0 (#261)
REVERT: 89c9597c91 Release 2.6.0 (#260)

git-subtree-dir: thirdParty/mallocMC
git-subtree-split: 8624e978e87644ee7c59f66a3c242aa969c01fc0

psychocoderHPC committed:
Merge commit '5bd457291716544ea91a2c4950a82f93c72c134f' into topic-updateToCpp20

psychocoderHPC committed:
CI: remove HIP (temporary)

Remove testing HIP because the container environment with ubuntu 20.04
is not C++20 ready.
libstdc++ is not supporting C++20 and using libc++ from clang is not
possible.

PrometheusPi reviewed (approved):
reviewed "in detail" together with @steindev 
Looks good
